src/core/egg-herbie.rkt

```diff
@@ -3,15 +3,64 @@
 (require egg-herbie)
 (require ffi/unsafe ffi/unsafe/define)
 (require "../syntax/rules.rkt" "../syntax/sugar.rkt" "../syntax/syntax.rkt" "../syntax/types.rkt"
-         "../common.rkt" "../errors.rkt" "../timeline.rkt")
+         "../common.rkt" "../errors.rkt" "../timeline.rkt" "../alternative.rkt")
 
 (module+ test (require rackunit))
 
 (provide make-egraph egraph-add-expr egraph-run-rules
          egraph-get-simplest egraph-get-variants
          egraph-get-proof egraph-is-unsound-detected
          rule->egg-rules expand-rules get-canon-rule-name
-         remove-rewrites)
+         remove-rewrites run-egg make-egg-descriptor
+         (struct-out proof-input) (struct-out egraph-input))
+
+(struct egraph-input (exprs rules num-variants terms iter-limit node-limit const-folding) #:transparent)
+(struct proof-input (start end) #:transparent)
+
+(define (make-egg-descriptor exprs rules num-variants [terms #f] #:iter-limit [iter-limit #f] #:node-limit [node-limit (*node-limit*)] [const-folding #f])
+  (egraph-input exprs rules num-variants terms iter-limit node-limit const-folding))
+
+;; TODO : Main entry point return (cons (list (list variant)) (list proof))
+(define (run-egg  input 
+                  proof-input 
+                  precompute? 
+                  #:roots [root-locs (make-list (length (egraph-input-exprs input)) '())])
+  ;; TODO : Make simplify match
+  ;; TODO : Make rr match
+  (let ([egg-graph (make-egraph)]) 
+      (define node-ids (map (curry egraph-add-expr egg-graph) (egraph-input-exprs input)))
+      (define iter-data (egraph-run-rules egg-graph (egraph-input-node-limit input) (egraph-input-rules input) node-ids precompute? #:limit (egraph-input-iter-limit input)))
+
+      ;; TODO : SORT THIS OUT
+      (when (egraph-is-unsound-detected egg-graph) 
+       (warn 'unsound-rules #:url "faq.html#unsound-rules"
+             "Unsound rule application detected in e-graph. Results may not be sound."))
+      
+      (define variants (if (egraph-input-num-variants input)
+                            (get-rr-variants egg-graph node-ids input)
+                            (get-simplify-variant egg-graph node-ids iter-data)))
+      (cond
+        [(proof-input? proof-input)
+          (begin
+            (define proof (egraph-get-proof egg-graph (proof-input-start proof-input) (proof-input-end proof-input)))
+            (when (null? proof)
+              (error (format "Failed to produce proof for ~a to ~a" (proof-input-start proof-input) (proof-input-end proof-input))))
+            (cons variants proof))]
+        [else (cons variants #f)])))
+
+(define (get-rr-variants egg-graph node-ids input)
+  (define variants
+        (for/list ([id node-ids] [expr (egraph-input-exprs input)]) ; TODO: Get Locations and Variants if Possible
+          (define output (egraph-get-variants egg-graph id expr))
+          (for/list ([variant (remove-duplicates output)])
+              (list variant input))))
+              variants)
```

Comment:
Please fix the indentation of `variants`, but also you don't need to define it if you're just going to return it immediately.